go_lib/hooks/generate_password/hook.go

```diff
@@ -17,53 +17,211 @@ limitations under the License.
+	BasicAuthPlainField string
+
+	FilterFunc func(secret *v1.Secret) (go_hook.FilterResult, error)
+}
+
+type Values struct {
+	ModuleKey           string
+	PasswordKey         string
+	PasswordInternalKey string
+	ExternalAuthKey     string
+}
+
+// Filter extracts password from the Secret. Password can be stored as a raw string or as
+// a basic auth plain format (user:{PLAIN}password). Custom FilterFunc is called for custom
+// password extraction.
+func (h *Hook) Filter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	secret := &v1.Secret{}
+	err := sdk.FromUnstructured(obj, secret)
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert secret to struct: %v", err)
 	}
 
-	if input.Values.Exists(fmt.Sprintf("%s.auth.password", moduleValuesPath)) {
-		return nil
+	if h.Secret.FilterFunc != nil {
+		return h.Secret.FilterFunc(secret)
 	}
 
-	if !input.ConfigValues.Exists(fmt.Sprintf("%s.auth", moduleValuesPath)) {
-		input.ConfigValues.Set(fmt.Sprintf("%s.auth", moduleValuesPath), json.RawMessage("{}"))
+	if h.Secret.RawPasswordField != "" {
+		return string(secret.Data[h.Secret.RawPasswordField]), nil
 	}
 
-	generatedPass := pwgen.AlphaNum(20)
+	// Return field with basic auth.
+	return string(secret.Data[h.SecretField()]), nil
+}
 
-	input.ConfigValues.Set(fmt.Sprintf("%s.auth.password", moduleValuesPath), generatedPass)
+// Handle restores password from the configuration or from the Secret and
+// puts it to internal values.
+// It generates new password if no password found in the configuration and the
+// Secret or no externalAuthentication defined.
+func (h *Hook) Handle(input *go_hook.HookInput) error {
+	var externalAuthKey = h.ExternalAuthKey()
+	var passwordKey = h.PasswordKey()
+	var passwordInternalKey = h.PasswordInternalKey()
+
+	// Clear password from internal values if an external authentication is enabled.
+	if input.Values.Exists(externalAuthKey) {
+		input.Values.Remove(passwordInternalKey)
+		return nil
+	}
 
-	return nil
-}
+	// Try to set password from config values.
+	password, ok := input.ConfigValues.GetOk(passwordKey)
+	if ok {
+		input.Values.Set(passwordInternalKey, password.String())
+		return nil
+	}
 
-func generatePassword(moduleValuesPath string) func(input *go_hook.HookInput) error {
-	return func(input *go_hook.HookInput) error {
-		err := generatePasswordWithArgs(input, moduleValuesPath)
+	// Try to restore password from the Secret.
+	snap := input.Snapshots[secretBindingName]
+	if len(snap) > 0 {
+		secretField, ok := snap[0].(string)
+		if !ok {
+			return fmt.Errorf("problem getting field '%s' from Secret/%s: got %T, while string is expected", h.SecretField(), h.Secret.Name, snap[0])
+		}
+		storedPassword, err := h.extractPasswordFromBasicAuth(secretField)
 		if err != nil {
 			return err
 		}
+		input.Values.Set(passwordInternalKey, storedPassword)
 		return nil
 	}
+
+	// No config value, no Secret, generate new password.
+	newPassword := ""
+	if h.PasswordGeneratorFunc != nil {
```

Comment:
Are there any cases of using different functions for the password generation?

Reply:
That was for testing, not needed.